mock/mock_proxy.go

```diff
@@ -0,0 +1,280 @@
+	proxy.ShimConnected <- true
+}
+
+func signalShimHandler(data []byte, userData interface{}, response *handlerResponse) {
+	client := userData.(*client)
+	proxy := client.proxy
+
+	signalPayload := ShimSignal{}
+	err := json.Unmarshal(data, &signalPayload)
+	assert.Nil(proxy.t, err)
+
+	proxy.logF("Proxy received signal: %v", signalPayload)
+
+	proxy.Signal <- signalPayload
+}
+
+func disconnectShimHandler(data []byte, userData interface{}, response *handlerResponse) {
+	client := userData.(*client)
+	proxy := client.proxy
+
+	proxy.log("Client sent DisconnectShim Command")
+	proxy.ShimDisconnected <- true
+}
+
+func stdinShimHandler(data []byte, userData interface{}, response *handlerResponse) {
+	client := userData.(*client)
+	proxy := client.proxy
+
+	proxy.lastStdinStream = data
+	proxy.StdinReceived <- true
+}
+
+func (proxy *MockProxy) SendStdoutStream(payload []byte) {
+	err := api.WriteStream(proxy.cl, api.StreamStdout, payload)
+	assert.Nil(proxy.t, err)
+}
+
+func (proxy *MockProxy) SendStderrStream(payload []byte) {
+	err := api.WriteStream(proxy.cl, api.StreamStderr, payload)
+	assert.Nil(proxy.t, err)
+}
+
+func (proxy *MockProxy) SendExitNotification(payload []byte) {
+	err := api.WriteNotification(proxy.cl, api.NotificationProcessExited, payload)
+	assert.Nil(proxy.t, err)
+}
+
+func (proxy *MockProxy) startListening() {
+
+	l, err := net.ListenUnix("unix", &net.UnixAddr{Name: proxy.connectionPath, Net: "unix"})
+	assert.Nil(proxy.t, err)
+
+	proxy.logF("listening on ", proxy.connectionPath)
+
+	proxy.listener = l
+}
+
+func (proxy *MockProxy) serveClient(proto *protocol, newConn net.Conn) {
+	newClient := &client{
+		proxy: proxy,
+		conn:  newConn,
+	}
+	err := proto.Serve(newConn, newClient)
+	proxy.logF("Error serving client : %v\n", err)
+
+	//Check that connection was closed and no other error occured.
+	//if err != nil {
+	//	assert.Equal(proxy.t, err, io.EOF)
+	//}
+
+	newConn.Close()
+	proxy.log("Client closed connection")
+
+	proxy.wg.Done()
+}
+
+func (proxy *MockProxy) serve() {
+	proto := newProtocol()
+
+	proto.Handle(FrameKey{api.TypeCommand, int(api.CmdConnectShim)}, connectShimHandler)
+	proto.Handle(FrameKey{api.TypeCommand, int(api.CmdDisconnectShim)}, disconnectShimHandler)
+	proto.Handle(FrameKey{api.TypeCommand, int(api.CmdSignal)}, signalShimHandler)
+	proto.Handle(FrameKey{api.TypeCommand, int(api.CmdConnectShim)}, connectShimHandler)
+
+	proto.Handle(FrameKey{api.TypeStream, int(api.StreamStdin)}, stdinShimHandler)
+
+	//Wait for a single client connection
+	conn, err := proxy.listener.Accept()
+	assert.Nil(proxy.t, err)
+	assert.NotNil(proxy.t, conn)
+	proxy.log("Client connected")
+
+	proxy.cl = conn
+
+	proxy.serveClient(proto, conn)
+}
+
+func (proxy *MockProxy) Start() {
+
+	proxy.startListening()
+	proxy.wg.Add(1)
+	go proxy.serve()
+}
+
+func (proxy *MockProxy) Stop() {
+
+	proxy.listener.Close()
+	if proxy.cl != nil {
+		proxy.log("Closing client connection")
+		proxy.cl.Close()
```

Comment:
I think it is safer if you assign proxy.cl to nil after you closed it. That way, we will never get back inside this statement.